source/websocket_encoder.c

```diff
@@ -0,0 +1,347 @@
+    AWS_FATAL_ASSERT(
+        (out_buf->buffer == prev_buf.buffer) && (out_buf->capacity == prev_buf.capacity) &&
+        (out_buf->len >= prev_buf.len));
+
+    size_t bytes_written = out_buf->len - prev_buf.len;
+
+    err = aws_add_u64_checked(encoder->state_bytes_processed, bytes_written, &encoder->state_bytes_processed);
+    if (err) {
+        return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+    }
+
+    /* Mask data, if necessary.
+     * RFC-6455 Section 5.3 Client-to-Server Masking
+     * Each byte of payload is XOR against a byte of the masking-key */
+    if (encoder->frame.masked) {
+        uint64_t mask_index = prev_bytes_processed;
+
+        /* Optimization idea: don't do this 1 byte at a time */
+        uint8_t *current_byte = out_buf->buffer + prev_buf.len;
+        uint8_t *end_byte = out_buf->buffer + out_buf->len;
+        while (current_byte != end_byte) {
+            *current_byte++ ^= encoder->frame.masking_key[mask_index++ % 4];
+        }
+    }
+
+    /* If done writing payload, proceed to next state */
+    if (encoder->state_bytes_processed == encoder->frame.payload_length) {
+        if (!user_says_done) {
+            return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+        }
+
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_DONE;
+    } else {
+        /* Some more error-checking... */
+        if (encoder->state_bytes_processed > encoder->frame.payload_length) {
+            return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+        }
+
+        if (user_says_done) {
+            return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static state_fn *s_state_functions[AWS_WEBSOCKET_ENCODER_STATE_DONE] = {
+    s_state_init,
+    s_state_opcode_byte,
+    s_state_length_byte,
+    s_state_extended_length,
+    s_state_masking_key_check,
+    s_state_masking_key,
+    s_state_payload_check,
+    s_state_payload,
+};
+
+int aws_websocket_encoder_process(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+
+    /* Run state machine until frame is completely decoded, or the state stops changing.
+     * Note that we don't necessarily stop looping when out_buf is full, because not all state need to output data */
+    while (encoder->state != AWS_WEBSOCKET_ENCODER_STATE_DONE) {
+        const enum aws_websocket_encoder_state prev_state = encoder->state;
+
+        int err = s_state_functions[encoder->state](encoder, out_buf);
+        if (err) {
+            return AWS_OP_ERR;
+        }
+
+        if (prev_state == encoder->state) {
+            /* This assert checks that each state is doing as much work as it possibly can.
+             * Except for the PAYLOAD state, where it's up to the user to fill the buffer. */
+            assert((out_buf->len == out_buf->capacity) || (encoder->state == AWS_WEBSOCKET_ENCODER_STATE_PAYLOAD));
```

Comment:
Is this just for you in dev? We'd rather not crash here later, right?

Reply:
it's a dev-assert